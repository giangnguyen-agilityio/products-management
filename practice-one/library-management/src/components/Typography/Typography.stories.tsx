import { type StoryFn, type Meta } from "@storybook/react";

import Typography from "./index";

export default {
  title: "Example/Typography",
  component: Typography,
  argTypes: {
    variant: {
      control: {
        type: "select",
        options: ["h1", "h2", "h3", "h4", "h5", "h6", "p"],
      },
    },
    color: {
      control: {
        type: "select",
        options: ["primary", "secondary"],
      },
    },
  },
} as Meta<typeof Typography>;

const Template: StoryFn<typeof Typography> = (args) => <Typography {...args} />;

export const Default = Template.bind({});
Default.args = {
  children: "This is a paragraph",
};
