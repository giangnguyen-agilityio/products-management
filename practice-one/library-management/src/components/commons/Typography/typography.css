.text-h1 {
  font-size: 32px;
}

.text-h2 {
  font-size: 24px;
}

.text-h3 {
  font-size: 20px;
}

.text-h4 {
  font-size: 18px;
  font-family: var(--secondary-font);
}

.text-h5 {
  font-size: 16px;
  font-family: var(--secondary-font);
}

.text-h6 {
  font-size: 14px;
  font-family: var(--secondary-font);
}

.text-p {
  font-size: 16px;
}

.text-primary {
  color: var(--primary-text-color);
}

.text-secondary {
  color: var(--secondary-text-color);
}
