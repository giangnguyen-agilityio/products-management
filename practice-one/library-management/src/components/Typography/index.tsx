import './typography.css'
import { memo } from 'react'

export interface TypographyProps {
  variant: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p'
  color?: 'primary' | 'secondary'
  className?: string
  children: React.ReactNode
}

const Typography = ({
  variant,
  color,
  children,
  className
}: TypographyProps) => {
  const textClasses = `text-${variant} text-${color ?? 'default'} ${
    className ?? ''
  }`

  switch (variant) {
    case 'h1':
      return <h1 className={textClasses}>{children}</h1>
    case 'h2':
      return <h2 className={textClasses}>{children}</h2>
    case 'h3':
      return <h3 className={textClasses}>{children}</h3>
    case 'h4':
      return <h4 className={textClasses}>{children}</h4>
    case 'h5':
      return <h5 className={textClasses}>{children}</h5>
    case 'h6':
      return <h6 className={textClasses}>{children}</h6>
    default:
      return <p className={textClasses}>{children}</p>
  }
}

export default memo(Typography)
