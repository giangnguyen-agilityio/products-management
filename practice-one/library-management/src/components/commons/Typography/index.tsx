import React, {memo} from 'react'
import './typography.css'

export interface TypographyProps {
  variant: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p' // The variant of the typography (heading or paragraph)
  color?: 'primary' | 'secondary' // The color of the text
  className?: string // Additional CSS class names
  children: React.ReactNode // The content of the typography
}

const Typography: React.FC<TypographyProps> = props => {
  const {variant, color, children, className} = props

  // Generate the CSS class names based on the props
  const textClasses = `text-${variant} text-${color ?? 'default'} ${
    className ?? ''
  }`

  // Render the typography based on the variant
  switch (variant) {
    case 'h1':
      return <h1 className={textClasses}>{children}</h1>
    case 'h2':
      return <h2 className={textClasses}>{children}</h2>
    case 'h3':
      return <h3 className={textClasses}>{children}</h3>
    case 'h4':
      return <h4 className={textClasses}>{children}</h4>
    case 'h5':
      return <h5 className={textClasses}>{children}</h5>
    case 'h6':
      return <h6 className={textClasses}>{children}</h6>
    default:
      return <p className={textClasses}>{children}</p>
  }
}

export default memo(Typography)
