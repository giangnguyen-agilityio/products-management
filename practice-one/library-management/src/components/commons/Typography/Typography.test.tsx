// Typography.test.tsx
import React from 'react'
import {render} from '@testing-library/react'
import Typography, {TypographyProps} from '.'

describe('Typography component', () => {
  const mockTypographyProps: TypographyProps = {
    variant: 'p',
    children: 'Test Typography',
  }

  it('should renders a paragraph with default styles when variant is not specified', () => {
    const {container, getByText} = render(
      <Typography {...mockTypographyProps} />
    )
    const textContent = getByText('Test Typography')
    expect(textContent.tagName.toLowerCase()).toBe('p')
    expect(textContent).toHaveClass('text-p')
    expect(container).toMatchSnapshot()
  })

  it('should renders different heading elements when variant is specified', () => {
    const variants: TypographyProps['variant'][] = [
      'h1',
      'h2',
      'h3',
      'h4',
      'h5',
      'h6',
    ]

    variants.forEach(variant => {
      const {container} = render(
        <Typography {...mockTypographyProps} variant={variant} />
      )
      const headingElement = container.querySelector(variant)

      expect(headingElement?.tagName).toBe(variant.toUpperCase())
      expect(container).toMatchSnapshot()
    })
  })

  it('should applies specified color class when color prop is provided', () => {
    const {getByText} = render(
      <Typography {...mockTypographyProps} color="primary" />
    )
    const textContent = getByText('Test Typography')
    expect(textContent).toHaveClass('text-primary')
  })

  it('should applies additional className when className prop is provided', () => {
    const {getByText} = render(
      <Typography {...mockTypographyProps} className="test-typography-class" />
    )
    const textContent = getByText('Test Typography')
    expect(textContent).toHaveClass('test-typography-class')
  })

  it('should applies both color and additional className when both props are provided', () => {
    const {getByText} = render(
      <Typography
        {...mockTypographyProps}
        color="primary"
        className="test-typography-class"
      />
    )
    const textContent = getByText('Test Typography')
    expect(textContent).toHaveClass('text-primary')
    expect(textContent).toHaveClass('test-typography-class')
  })
})
