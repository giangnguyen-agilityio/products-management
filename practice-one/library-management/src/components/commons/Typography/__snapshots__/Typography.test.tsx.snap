// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography component should renders a paragraph with default styles when variant is not specified 1`] = `
<div>
  <p
    class="text-p text-default "
  >
    Test Typography
  </p>
</div>
`;

exports[`Typography component should renders different heading elements when variant is specified 1`] = `
<div>
  <h1
    class="text-h1 text-default "
  >
    Test Typography
  </h1>
</div>
`;

exports[`Typography component should renders different heading elements when variant is specified 2`] = `
<div>
  <h2
    class="text-h2 text-default "
  >
    Test Typography
  </h2>
</div>
`;

exports[`Typography component should renders different heading elements when variant is specified 3`] = `
<div>
  <h3
    class="text-h3 text-default "
  >
    Test Typography
  </h3>
</div>
`;

exports[`Typography component should renders different heading elements when variant is specified 4`] = `
<div>
  <h4
    class="text-h4 text-default "
  >
    Test Typography
  </h4>
</div>
`;

exports[`Typography component should renders different heading elements when variant is specified 5`] = `
<div>
  <h5
    class="text-h5 text-default "
  >
    Test Typography
  </h5>
</div>
`;

exports[`Typography component should renders different heading elements when variant is specified 6`] = `
<div>
  <h6
    class="text-h6 text-default "
  >
    Test Typography
  </h6>
</div>
`;
